File: my-reader-mobile/modules/readium/ios/Reader/EPUB/ReaderNoteMarkerDecoration.swift
import ReadiumNavigator
import UIKit

let readerNoteMarkerStyleId: Decoration.Style.Id = "myreader-note-marker"

struct ReaderNoteMarkerStyleConfig: Hashable {
  let tint: UIColor
}

func readerNoteMarkerStyle(tint: String?) -> Decoration.Style {
  Decoration.Style(
    id: readerNoteMarkerStyleId,
    config: ReaderNoteMarkerStyleConfig(
      tint: tint.flatMap(UIColor.fromCSS) ?? UIColor(red: 0.85, green: 0.66, blue: 0.16, alpha: 1)
    )
  )
}

func readerNoteMarkerTemplate() -> HTMLDecorationTemplate {
  HTMLDecorationTemplate(
    layout: .boxes,
    width: .wrap,
    element: { decoration in
      let config = decoration.style.config as? ReaderNoteMarkerStyleConfig
      let tint = config?.tint.cssHex ?? "#D9A928"
      let accessibilityLabel =
        (decoration.userInfo["accessibilityLabel"] as? String ?? "Open note")
          .readerHTMLAttributeEscaped
      return ReaderNoteMarkerTemplateSource.element
        .replacingOccurrences(of: "{{accessibilityLabel}}", with: accessibilityLabel)
        .replacingOccurrences(of: "{{tint}}", with: tint)
    },
    stylesheet: ReaderNoteMarkerTemplateSource.stylesheet
      .replacingOccurrences(of: "{{hitSize}}", with: "44")
      .replacingOccurrences(of: "{{hitOffset}}", with: "22")
  )
}

private extension String {
  var readerHTMLAttributeEscaped: String {
    replacingOccurrences(of: "&", with: "&amp;")
      .replacingOccurrences(of: "\"", with: "&quot;")
      .replacingOccurrences(of: "<", with: "&lt;")
      .replacingOccurrences(of: ">", with: "&gt;")
  }
}

File: my-reader-mobile/modules/readium/android/src/main/java/com/myreader/readium/reader/ReaderNoteMarkerDecoration.kt
package com.myreader.readium.reader

import android.graphics.Color
import androidx.annotation.ColorInt
import kotlinx.parcelize.Parcelize
import org.readium.r2.navigator.Decoration
import org.readium.r2.navigator.html.HtmlDecorationTemplate
import org.readium.r2.navigator.html.HtmlDecorationTemplates

@Parcelize
internal data class ReaderNoteMarkerDecorationStyle(
  @ColorInt override val tint: Int,
) : Decoration.Style, Decoration.Style.Tinted

internal fun readerDecorationTemplates(): HtmlDecorationTemplates =
  HtmlDecorationTemplates.defaultTemplates().apply {
    set(ReaderNoteMarkerDecorationStyle::class, readerNoteMarkerTemplate())
  }

private fun readerNoteMarkerTemplate(): HtmlDecorationTemplate =
  HtmlDecorationTemplate(
    layout = HtmlDecorationTemplate.Layout.BOXES,
    width = HtmlDecorationTemplate.Width.WRAP,
    element = { decoration ->
      val tint = (decoration.style as? ReaderNoteMarkerDecorationStyle)?.tint
        ?: Color.rgb(217, 169, 40)
      val accessibilityLabel =
        decoration.extras["accessibilityLabel"]?.toString()?.htmlAttributeEscaped()
          ?: "Open note"
      ReaderNoteMarkerTemplateSource.ELEMENT
        .replace("{{accessibilityLabel}}", accessibilityLabel)
        .replace("{{tint}}", tint.toCssColor())
    },
    stylesheet =
      ReaderNoteMarkerTemplateSource.STYLESHEET
        .replace("{{hitSize}}", "48")
        .replace("{{hitOffset}}", "24"),
  )

private fun Int.toCssColor(): String =
  "rgba(${Color.red(this)}, ${Color.green(this)}, ${Color.blue(this)}, ${Color.alpha(this) / 255.0})"

private fun String.htmlAttributeEscaped(): String =
  replace("&", "&amp;")
    .replace("\"", "&quot;")
    .replace("<", "&lt;")
    .replace(">", "&gt;")
